lib/npg_pipeline/function/s3_archiver.pm

```diff
@@ -0,0 +1,265 @@
+
+Readonly::Scalar my $CONFIG_ITEMS_KEY   => 's3_archive';
+Readonly::Scalar my $CONFIG_STUDY_KEY   => 'study_id';
+Readonly::Scalar my $CONFIG_URL_KEY     => 'url';
+
+our $VERSION = '0';
+
+
+=head2 expected_files
+
+  Arg [1]    : Data product whose files to list,npg_pipeline::product.
+
+  Example    : my @files = $obj->expected_files($product)
+  Description: Return a list of the files expected to to present for
+               archiving in the runfolder.
+
+  Returntype : Array
+
+=cut
+
+sub expected_files {
+  my ($self, $product) = @_;
+
+  $product or $self->logconfess('A product argument is required');
+
+  my @expected_files;
+
+  my $dir_path = catdir($self->archive_path(), $product->dir_path());
+  my @extensions = qw{cram cram.md5 cram.crai
+                      seqchksum sha512primesums512.seqchksum
+                      F0x900.stats F0xB00.stats F0xF04_target.stats
+                      bcfstats};
+  push @expected_files,
+    map { $product->file_path($dir_path, ext => $_) } @extensions;
+
+  my $qc_path = $product->qc_out_path($self->archive_path());
+
+  my @qc_extensions = qw{verify_bam_id.json};
+  push @expected_files,
+    map { $product->file_path($qc_path, ext => $_) } @qc_extensions;
+
+  @expected_files = sort @expected_files;
+
+  return @expected_files;
+}
+
+=head2 create
+
+  Arg [1]    : None
+
+  Example    : my $defs = $obj->create
+  Description: Create per-product data file function definitions.
+
+  Returntype : ArrayRef[npg_pipeline::function::definition]
+
+=cut
+
+sub create {
+  my ($self) = @_;
+
+  my $id_run         = $self->id_run();
+  my $archive_config = $self->_read_config();
+
+  my @definitions;
+
+  my $i = 0;
+  foreach my $product (@{$self->products->{data_products}}) {
+    if ($product->is_tag_zero_product) {
+      $self->info('Skipping archiving for tag zero product ',
+                  $product->file_name_root);
+      next;
+    }
+    if ($product->lims->is_control) {
+      $self->info('Skipping archiving for control product ',
+                  $product->file_name_root);
+      next;
+    }
+
+    my $sample     = $product->lims->sample_supplier_name;
+    $sample or $self->logcroak('Failed to get a supplier sample name',
+                               'for product ', $product->file_name_root);
+
+    my $study_id   = $product->lims->study_id;
+    $study_id or $self->logcroak('Failed to get a study_id',
+                                 'for product ', $product->file_name_root);
+
+    if (exists $archive_config->{$study_id}) {
+      $self->info(sprintf q{S3 archiving %s in study %s to bucket URL %s},
+                  $product->file_name_root, $study_id,
+                  $archive_config->{$study_id});
+    } else {
+      $self->info(sprintf q{Skipping S3 archiving %s in study %s},
+                  $product->file_name_root, $study_id);
+      next;
+    }
+
+    my $job_name = sprintf q{%s_%d_%d}, $ARCHIVE_EXECUTABLE, $id_run, $i;
```

Comment:
There is no need to index job names, for LSF it might be even wrong. It might be benefitial to use a more descriptive job name, say, copy_data2aws or aws_archival